modules/dataquery/php/endpoints/queries/query/run.class.inc

```diff
@@ -57,14 +63,44 @@ class Run extends \LORIS\Http\Endpoint
         try {
             $query = new \LORIS\dataquery\Query($this->loris, $queryID);
 
+            $db = $this->loris->getDatabaseConnection();
+            // Put the query run in a transaction to make it atomic
+            $db->beginTransaction();
             $queryRun = $query->newRun($user);
-            $queryRun->insertCandidates($user);
-            $data = $queryRun->getQueryDataProvisioner();
+            $data     = $queryRun->getQueryDataProvisioner()->execute($user);
+
+            $values = [];
+
+            set_time_limit(300);
+            $runid  = $queryRun->getRunId();
+            $stream = new \LORIS\Http\DataIteratorBinaryStream(
+                $data,
+                function (
+                    $candid,
+                    $rowsArray,
+                    /* @phan-suppress-next-line PhanUnusedClosureParameter */
```

Comment:
Closure signature needs to match the callback of DataIteratorBinaryStream, so the unused parameter can't be removed.